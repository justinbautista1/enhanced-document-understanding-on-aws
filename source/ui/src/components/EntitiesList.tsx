// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import {
    Badge,
    Box,
    Button,
    Checkbox,
    ExpandableSection,
    SpaceBetween,
    StatusIndicator,
    StatusIndicatorProps,
    Tabs,
    Toggle
} from '@cloudscape-design/components';
import React, { useCallback } from 'react';
import { useAppDispatch, useAppSelector } from '../store/hooks/hooks';
import {
    useLazyDocumentToDownloadQuery,
    useLazyGetDocumentByCaseAndDocumentIdQuery
} from '../store/reducers/documentApiSlice';
import { selectEntityStatus, setStatus } from '../store/reducers/entitySlice';
import { useRedactMutation } from '../store/reducers/redactApiSlice';
import { isStatusSuccess, renderStatus } from '../utils/common-renderers';
import { COMPREHEND_MEDICAL_SERVICE, ENTITIES, EntityTypes, PREVIEW_REDACTION_ON } from '../utils/constants';
import './EntitiesList.css';
import { getEntitiesToProcess } from './entityUtils';

type EntitiesListProps = {
    entities: any;
    documentPageCount: number;
    currentPageNumber: number;
    switchPage: Function;
    comprehendService: string;
    entityType: string;
    standardEntities: any;
    medicalEntities: any;
    piiEntities: any;
    selectedEntities: any;
    setSelectedEntities: Function;
    selectedDocumentId: string | null;
    selectedCaseId: string | null;
    previewRedaction: string;
    setPreviewRedaction: Function;
    currentStatus: StatusIndicatorProps.Type | undefined;
    phrase: string;
    setPhrase: Function;
    inputPhrase: string;
    setInputPhrase: Function;
};

type EntityItemProps = {
    entityKey: string;
    selectedEntities: any;
    handleEntitySelect: (entityPath: string[]) => void;
    switchPage: Function;
    entityPath: string[];
    entityType: string;
    entityObject: any;
};

const EntityItem: React.FC<EntityItemProps> = ({
    entityKey,
    selectedEntities,
    handleEntitySelect,
    switchPage,
    entityPath,
    entityType,
    entityObject
}) => {
    const handleCheckboxChange = (entityPath: string[]) => {
        handleEntitySelect(entityPath);
    };

    return (
        <div>
            <SpaceBetween direction="horizontal" size="xs">
                <Box data-testid="box-view-entity" padding={{ top: 'xxs' }} margin={{ top: 'xxxs' }}>
                    <Checkbox
                        checked={
                            selectedEntities[entityType]
                                ? selectedEntities[entityType].some((subArray: string[]) => {
                                      return (
                                          subArray.length === entityPath.length &&
                                          subArray.every((value, index) => value === entityPath[index])
                                      );
                                  })
                                : false
                        }
                        onChange={() => handleCheckboxChange(entityPath)}
                    ></Checkbox>
                </Box>
                <Box display="inline">
                    <ExpandableSection
                        headerText={entityKey}
                        headingTagOverride="h5"
                        data-testid={'entities-list-expandable-' + entityKey}
                    >
                        {Object.keys(entityObject).map((page) => (
                            <Box data-testid="box-view-entity-page" key={entityKey + ' ' + page + '-key'}>
                                <SpaceBetween direction="vertical" size="xs">
                                    <SpaceBetween direction="horizontal" size="xs">
                                        <Checkbox
                                            checked={
                                                selectedEntities[entityType]
                                                    ? selectedEntities[entityType].some((subArray: string[]) => {
                                                          return (
                                                              subArray.length === entityPath.length + 1 &&
                                                              subArray.every(
                                                                  (value, index) =>
                                                                      value === [...entityPath, page][index]
                                                              )
                                                          );
                                                      })
                                                    : false
                                            }
                                            onChange={() => handleCheckboxChange([...entityPath, page])}
                                        ></Checkbox>
                                        <span onClick={() => switchPage(+page)}>{'Page ' + page}</span>
                                    </SpaceBetween>
                                    {/* Per-instance checkboxes */}
                                    {Array.isArray(entityObject[page]) &&
                                        entityObject[page].map((instance: any, idx: number) => {
                                            // Extract a snippet: up to 5 words from the instance text
                                            let snippet = '';
                                            if (instance.Text) {
                                                snippet = instance.Text.split(/\s+/).slice(0, 5).join(' ');
                                            } else if (typeof instance === 'string') {
                                                snippet = instance.split(/\s+/).slice(0, 5).join(' ');
                                            } else {
                                                snippet = 'Instance ' + (idx + 1);
                                            }
                                            return (
                                                <SpaceBetween
                                                    direction="horizontal"
                                                    size="xs"
                                                    key={entityKey + '-' + page + '-instance-' + idx}
                                                >
                                                    <Checkbox
                                                        checked={
                                                            selectedEntities[entityType]
                                                                ? selectedEntities[entityType].some(
                                                                      (subArray: string[]) => {
                                                                          return (
                                                                              subArray.length ===
                                                                                  entityPath.length + 2 &&
                                                                              subArray.every(
                                                                                  (value, index) =>
                                                                                      value ===
                                                                                      [
                                                                                          ...entityPath,
                                                                                          page,
                                                                                          idx.toString()
                                                                                      ][index]
                                                                              )
                                                                          );
                                                                      }
                                                                  )
                                                                : false
                                                        }
                                                        onChange={() =>
                                                            handleCheckboxChange([...entityPath, page, idx.toString()])
                                                        }
                                                    ></Checkbox>
                                                    <span
                                                        title={
                                                            instance.Text ||
                                                            (typeof instance === 'string' ? instance : undefined)
                                                        }
                                                    >
                                                        {snippet}
                                                    </span>
                                                </SpaceBetween>
                                            );
                                        })}
                                </SpaceBetween>
                            </Box>
                        ))}
                    </ExpandableSection>
                </Box>
            </SpaceBetween>
        </div>
    );
};

const EntitiesList: React.FC<EntitiesListProps> = (props) => {
    const dispatch = useAppDispatch();
    const [redact] = useRedactMutation();

    const [getDocumentByCaseAndDocumentIdTrigger] = useLazyGetDocumentByCaseAndDocumentIdQuery();

    const [getDocumentToDownloadTrigger] = useLazyDocumentToDownloadQuery();

    const currentStatus = useAppSelector(selectEntityStatus);

    const getFilteredArray = useCallback(
        (entityType: string) => {
            return props.selectedEntities[entityType].filter((item: string[]) => {
                if (
                    props.selectedEntities[entityType].some(
                        (otherItem: string[]) =>
                            otherItem.length < item.length &&
                            otherItem.slice(0, item.length - 1).every((val, index) => val === item[index])
                    )
                ) {
                    return false;
                }

                return true;
            });
        },
        [props.selectedEntities]
    );

    const concatPages = useCallback(
        (entities: any, entitiesToRedact: any, selectedEntity: string[], entity: string, entityType: string) => {
            for (const page of Object.keys(entities[selectedEntity[0]][entity])) {
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][entity] = entitiesToRedact[ENTITIES][
                    entityType
                ][selectedEntity[0]][entity].concat(+page);
            }
            return entitiesToRedact;
        },
        []
    );

    const addEntityTypeToRedact = useCallback(
        (entities: any, entitiesToRedact: any, selectedEntity: string[], entityType: string) => {
            entitiesToRedact[ENTITIES][entityType][selectedEntity[0]] = {};
            for (const entity of Object.keys(entities[selectedEntity[0]])) {
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][entity] = [];
                for (const page of Object.keys(entities[selectedEntity[0]][entity])) {
                    entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][entity] = entitiesToRedact[ENTITIES][
                        entityType
                    ][selectedEntity[0]][entity].concat(+page);
                }
            }
            return entitiesToRedact;
        },
        []
    );

    const addEntityValueToRedact = useCallback(
        (entities: any, entitiesToRedact: any, selectedEntity: string[], entityType: string) => {
            if (entitiesToRedact[ENTITIES][entityType].hasOwnProperty(selectedEntity[0])) {
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][selectedEntity[1]] = [];
            } else {
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]] = {};
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][selectedEntity[1]] = [];
            }
            return concatPages(entities, entitiesToRedact, selectedEntity, selectedEntity[1], entityType);
        },
        [concatPages]
    );

    const addEntityPageToRedact = useCallback(
        (entities: any, entitiesToRedact: any, selectedEntity: string[], entityType: string) => {
            if (entitiesToRedact[ENTITIES][entityType].hasOwnProperty(selectedEntity[0])) {
                if (entitiesToRedact[ENTITIES][entityType][selectedEntity[0]].hasOwnProperty(selectedEntity[1])) {
                    entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][selectedEntity[1]] = entitiesToRedact[
                        ENTITIES
                    ][entityType][selectedEntity[0]][selectedEntity[1]].concat(+selectedEntity[2]);
                } else {
                    entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][selectedEntity[1]] = [+selectedEntity[2]];
                }
            } else {
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]] = {};
                entitiesToRedact[ENTITIES][entityType][selectedEntity[0]][selectedEntity[1]] = [+selectedEntity[2]];
            }
            return entitiesToRedact;
        },
        []
    );

    const redactAllEntities = async () => {
        try {
            dispatch(setStatus('loading'));
            let entitiesToRedact: any = {
                'entities': {
                    [EntityTypes.ENTITY_STANDARD]: {},
                    [EntityTypes.MEDICAL_ENTITY]: {},
                    [EntityTypes.PII]: {}
                }
            };

            for (const entityType of Object.values(EntityTypes)) {
                if (props.selectedEntities[entityType] && props.selectedEntities[entityType].length > 0) {
                    const entities = getEntitiesToProcess(entityType, props);
                    const filteredArray = getFilteredArray(entityType);
                    for (const selectedEntity of filteredArray) {
                        entitiesToRedact = processSelectedEntity(
                            selectedEntity,
                            entitiesToRedact,
                            entities,
                            entityType
                        );
                    }
                } else {
                    delete entitiesToRedact.entities[entityType];
                }
            }

            const phrasePages = props?.entities?.OTHER?.[props.phrase];
            const phrasesToRedact = {
                phrases: [
                    {
                        text: props.phrase,
                        pages: Object.keys(phrasePages).map((page) => +page)
                    }
                ]
            };

            const textsToRedact = Object.assign({}, entitiesToRedact, phrasesToRedact);
            console.log('textsToRedact', textsToRedact);
            console.log('entitiesToRedact', entitiesToRedact);

            await redact({
                caseId: props.selectedCaseId,
                documentId: props.selectedDocumentId,
                body: textsToRedact
            }).unwrap();

            const documentResponse = await getDocumentByCaseAndDocumentIdTrigger({
                caseId: props.selectedCaseId as string,
                documentId: props.selectedDocumentId as string,
                redacted: true
            }).unwrap();

            const signedUrlObject = await getDocumentToDownloadTrigger({ key: documentResponse.key }).unwrap();

            window.open(signedUrlObject.downloadUrl, '_blank', 'noopener');
            dispatch(setStatus('success'));
        } catch (err) {
            dispatch(setStatus('error'));
            console.log(err);
            console.error(err);
        }

        function processSelectedEntity(selectedEntity: any, entitiesToRedact: any, entities: any, entityType: string) {
            switch (selectedEntity.length) {
                case 1:
                    entitiesToRedact = addEntityTypeToRedact(entities, entitiesToRedact, selectedEntity, entityType);
                    break;
                case 2:
                    entitiesToRedact = addEntityValueToRedact(entities, entitiesToRedact, selectedEntity, entityType);
                    break;
                case 3:
                    entitiesToRedact = addEntityPageToRedact(entities, entitiesToRedact, selectedEntity, entityType);
            }
            return entitiesToRedact;
        }
    };

    let is_comprehend_medical = '';
    let medical_test_id = '';
    if (props.comprehendService === COMPREHEND_MEDICAL_SERVICE) {
        is_comprehend_medical = 'Medical';
        medical_test_id = 'medical-';
    }

    // when success message is received and no data
    if (!props.entities || !Object.keys(props.entities).length) {
        if (isStatusSuccess(props.currentStatus))
            return (
                <p data-testid={`${medical_test_id}entities-nodata`}>
                    No {`${is_comprehend_medical}`} Entities detected
                </p>
            );
    }

    const inSelectedEntities = (entityPath: string[]) => {
        return props.selectedEntities[props.entityType].some((subArray: string[]) => {
            return (
                subArray.length === entityPath.length && subArray.every((value, index) => value === entityPath[index])
            );
        });
    };

    const removeFromSelectedEntities = (entityPath: string[]) => {
        props.setSelectedEntities({
            ...props.selectedEntities,
            [props.entityType]: props.selectedEntities[props.entityType].filter((arr: string[]) => {
                if (arr.slice(0, entityPath.length).every((elem, i) => elem === entityPath[i])) {
                    return false;
                }
                return true;
            })
        });
    };

    const handleEntitySelect = (entityPath: string[]) => {
        if (inSelectedEntities(entityPath)) {
            removeFromSelectedEntities(entityPath);
        } else {
            let newSelectedEntities: string[][] = [entityPath];
            if (entityPath.length === 2) {
                for (const pageNum of Object.keys(props.entities[entityPath[0]][entityPath[1]])) {
                    newSelectedEntities.push([...entityPath, pageNum]);
                }
            } else if (entityPath.length === 1) {
                for (const entityName of Object.keys(props.entities[entityPath[0]])) {
                    newSelectedEntities.push([...entityPath, entityName]);
                    for (const pageNum of Object.keys(props.entities[entityPath[0]][entityName])) {
                        newSelectedEntities.push([...entityPath, entityName, pageNum]);
                    }
                }
            }

            props.setSelectedEntities({
                ...props.selectedEntities,
                [props.entityType]: [...props.selectedEntities[props.entityType], ...newSelectedEntities]
            });
        }
    };

    const selectAllEntities = () => {
        let allSelectedEntities: string[][] = [];
        for (const entityType of Object.keys(props.entities)) {
            allSelectedEntities.push([entityType]);
            for (const entityName of Object.keys(props.entities[entityType])) {
                allSelectedEntities.push([entityType, entityName]);
                for (const pageNum of Object.keys(props.entities[entityType][entityName])) {
                    allSelectedEntities.push([entityType, entityName, pageNum]);
                }
            }
        }
        props.setSelectedEntities({
            ...props.selectedEntities,
            [props.entityType]: allSelectedEntities
        });
    };

    const deselectAllEntities = () => {
        props.setSelectedEntities({ ...props.selectedEntities, [props.entityType]: [] });
    };

    const mainTabs =
        props.entities &&
        Object.keys(props.entities).map((type) => ({
            label: (
                <div>
                    {type}{' '}
                    <Badge color="blue" data-testid="badge-entity-count">
                        {props.selectedEntities[props.entityType]
                            ? props.selectedEntities[props.entityType].filter(
                                  (subarray: string[]) => subarray[0] === type && subarray.length === 3
                              ).length
                            : 0}
                    </Badge>
                </div>
            ),
            id: type.replace(/\s/g, ''),
            content: (
                <Box data-testid="box-parent-view-entity-title">
                    {renderStatus(props.currentStatus, true, false, 'An error occurred loading detected entities.', '')}
                    {isStatusSuccess(props.currentStatus) && (
                        <SpaceBetween direction="horizontal" size="xs">
                            <Checkbox
                                checked={
                                    props.selectedEntities[props.entityType]
                                        ? props.selectedEntities[props.entityType].some((subArray: string[]) => {
                                              return (
                                                  subArray.length === 1 &&
                                                  subArray.every((value, index) => value === type)
                                              );
                                          })
                                        : false
                                }
                                onChange={() => handleEntitySelect([type])}
                            ></Checkbox>
                            <Box variant="h4" display="inline">
                                All Detected {type} Entities
                            </Box>
                        </SpaceBetween>
                    )}

                    {isStatusSuccess(props.currentStatus) &&
                        Object.keys(props.entities[type]).map((entity) => (
                            <EntityItem
                                key={type + ' ' + entity + '-key'}
                                entityType={props.entityType}
                                entityObject={props.entities[type][entity]}
                                entityKey={entity}
                                selectedEntities={props.selectedEntities}
                                handleEntitySelect={handleEntitySelect}
                                switchPage={props.switchPage}
                                entityPath={[type, entity]}
                            />
                        ))}
                </Box>
            )
        }));

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        const standard = props.selectedEntities['entity-standard'];
        const updatedStandard = standard.filter((entity: string[]) => entity[1] !== props.phrase);
        let newStandard = [...updatedStandard];
        if (props.inputPhrase) {
            newStandard.push(['OTHER', props.inputPhrase]);
            props.setPhrase(props.inputPhrase);
        }
        props.setSelectedEntities({
            ...props.selectedEntities,
            ['entity-standard']: newStandard
        });
    };

    const status = renderStatus(
        props.currentStatus,
        true,
        false,
        `An error occurred loading ${is_comprehend_medical} Entities.`,
        ''
    );
    if (isStatusSuccess(props.currentStatus)) {
        return (
            <div>
                <SpaceBetween size={'xs'} direction="horizontal">
                    <Button onClick={() => selectAllEntities()} data-testid="select-all-entities">
                        Select All
                    </Button>
                    <Button onClick={() => deselectAllEntities()} data-testid="deselect-all-entities">
                        Deselect All
                    </Button>
                    <Button
                        onClick={() => redactAllEntities()}
                        data-testid="redact-all-entities"
                        disabled={Object.values(props.selectedEntities as string[][]).every(
                            (subArray) => subArray.length === 0
                        )}
                    >
                        Download Redacted Document &nbsp;
                        {currentStatus && <StatusIndicator data-testid="status" type={currentStatus} />}
                    </Button>
                    <Toggle
                        onChange={({ detail }) => props.setPreviewRedaction(detail.checked ? PREVIEW_REDACTION_ON : '')}
                        checked={props.previewRedaction === PREVIEW_REDACTION_ON}
                        data-testid="toggle-preview-redaction"
                    >
                        {' '}
                        Preview Redacted Document{' '}
                    </Toggle>
                </SpaceBetween>
                <Tabs tabs={mainTabs}></Tabs>
                <form onSubmit={handleSubmit}>
                    <textarea
                        rows={10}
                        cols={50}
                        value={props.inputPhrase}
                        onChange={(e) => props.setInputPhrase(e.target.value)}
                    />
                    <button type="submit">Submit</button>
                </form>
            </div>
        );
    }
    return <p data-testid={`${medical_test_id}entities-status-only`}> {status} </p>;
};

export default EntitiesList;
