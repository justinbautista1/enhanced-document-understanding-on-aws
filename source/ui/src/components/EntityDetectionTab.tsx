// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Box, Container, StatusIndicatorProps } from '@cloudscape-design/components';
import { useCallback, useMemo, useState } from 'react';
import { renderStatus } from '../utils/common-renderers';
import { BoundingBox } from '../utils/interfaces';

import { EntityTypes } from '../utils/constants';
import DocumentRenderer from './DocumentRenderer/DocumentRenderer';
import EntitiesList from './EntitiesList';
import { getEntitiesToProcess } from './entityUtils';

type EntityDetectionTabProps = {
    selectedDocumentFileType: string | null;
    selectedDocumentUrl: string | null;
    selected?: string;
    standardEntities: any;
    medicalEntities: any;
    piiEntities: any;
    documentPageCount: number;
    currentPageNumber: number;
    switchPage: Function;
    comprehendService: string;
    entityType: string;
    selectedDocumentId: string | null;
    selectedCaseId: string | null;
    selectedEntities: any;
    setSelectedEntities: Function;
    previewRedaction: string;
    setPreviewRedaction: Function;
    currentStatus: StatusIndicatorProps.Type | undefined;
    dataTestId?: string;
    retrieveSignedUrl: Function;
    textractText: any;
    phrase: string;
    setPhrase: Function;
};

/**
 * EntityDetectionTab is used to render data for Generic, PII and Medical Comprehend inferences
 * @param props to be used to populate the tab
 * @returns
 */
export default function EntityDetectionTab(props: EntityDetectionTabProps) {
    let documentEntities = getEntitiesToProcess(props.entityType, props);
    const [inputPhrase, setInputPhrase] = useState<string>('');
    console.log('EntityDetectionTab documentEntities', documentEntities);
    console.log('selectedEntities', props.selectedEntities);

    const getFilteredArray = useCallback(
        (entityType: string) => {
            // Now allow selection at the instance level (4th element: instance index or id)
            return props.selectedEntities[entityType].filter((item: string[]) => {
                // If item has 4 elements, it's an instance selection: [entityType, entityValue, pageNumber, instanceIndex]
                if (item.length === 4) {
                    // Only show if current page matches
                    return item[2] === props.currentPageNumber.toString();
                }
                // If item has 3 elements, it's a page selection (legacy)
                if (item.length === 3) {
                    return item[2] === props.currentPageNumber.toString();
                }
                // If item has less, fallback to previous logic
                if (
                    props.selectedEntities[entityType].some(
                        (otherItem: string[]) =>
                            otherItem.length < item.length &&
                            otherItem.slice(0, item.length - 1).every((val, index) => val === item[index])
                    )
                ) {
                    return false;
                }
                return true;
            });
        },
        [props.currentPageNumber, props.selectedEntities]
    );

    // New: Get bounding box for a specific instance
    const getBoundingBoxForEntityInstance = useCallback((entities: any, entityPath: string[]) => {
        // entityPath: [entityType, entityValue, pageNumber, instanceIndex]
        let boundingBoxes: BoundingBox[] = [];
        const [type, value, page, instanceIdx] = entityPath;
        const pageInstances = entities[type]?.[value]?.[page];
        if (pageInstances && pageInstances[instanceIdx]) {
            boundingBoxes = boundingBoxes.concat(pageInstances[instanceIdx].BoundingBoxes);
        }
        return boundingBoxes;
    }, []);

    const getBoundingBoxesForEntityPage = useCallback((entities: any, entityPath: string[]) => {
        let boundingBoxes: BoundingBox[] = [];
        for (const pageInstance of entities[entityPath[0]][entityPath[1]][entityPath[2]]) {
            boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
        }
        return boundingBoxes;
    }, []);

    const getBoundingBoxesForEntityValue = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            if (props.currentPageNumber.toString() in entities[entityPath[0]][entityPath[1]]) {
                for (const pageInstance of entities[entityPath[0]][entityPath[1]][props.currentPageNumber.toString()]) {
                    boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const getBoundingBoxesForEntityType = useCallback(
        (entities: any, entityPath: string[]) => {
            let boundingBoxes: BoundingBox[] = [];
            for (const entity of Object.keys(entities[entityPath[0]])) {
                if (props.currentPageNumber.toString() in entities[entityPath[0]][entity]) {
                    for (const pageInstance of entities[entityPath[0]][entity][props.currentPageNumber.toString()]) {
                        boundingBoxes = boundingBoxes.concat(pageInstance.BoundingBoxes);
                    }
                }
            }
            return boundingBoxes;
        },
        [props.currentPageNumber]
    );

    const pageEntities = useMemo(() => {
        let boundingBoxes: BoundingBox[] = [];
        for (const entityType of Object.values(EntityTypes)) {
            if (props.selectedEntities[entityType] && props.selectedEntities[entityType].length > 0) {
                const entities = getEntitiesToProcess(entityType, props);
                const filteredArray = getFilteredArray(entityType);
                for (const entityPath of filteredArray) {
                    switch (entityPath.length) {
                        case 4:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxForEntityInstance(entities, entityPath));
                            break;
                        case 3:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityPage(entities, entityPath));
                            break;
                        case 2:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityValue(entities, entityPath));
                            break;
                        case 1:
                        default:
                            boundingBoxes = boundingBoxes.concat(getBoundingBoxesForEntityType(entities, entityPath));
                            break;
                    }
                }
            }
        }
        return boundingBoxes;
    }, [
        getBoundingBoxesForEntityPage,
        getBoundingBoxesForEntityType,
        getBoundingBoxesForEntityValue,
        getBoundingBoxForEntityInstance,
        getFilteredArray,
        props
    ]);

    const status = renderStatus(
        props.currentStatus,
        true,
        false,
        `An error occurred loading the document preview.`,
        ''
    );
    return (
        <div style={{ display: 'flex', height: '100%' }} data-testid={props.dataTestId}>
            <div style={{ width: '50%', float: 'left', paddingRight: '0.5%', paddingLeft: '1%' }}>
                <Container data-testid="document-rendering-container">
                    {status}
                    <DocumentRenderer
                        selectedDocumentFileType={props.selectedDocumentFileType}
                        selectedDocumentUrl={props.selectedDocumentUrl}
                        currentPageNumber={props.currentPageNumber}
                        switchPage={props.switchPage}
                        marks={pageEntities}
                        previewRedaction={props.previewRedaction}
                        retrieveSignedUrl={props.retrieveSignedUrl}
                        inputPhrase={inputPhrase}
                        setInputPhrase={setInputPhrase}
                    />
                </Container>
            </div>
            <div
                style={{
                    width: '50%',
                    float: 'left',
                    paddingLeft: '0.5%',
                    paddingRight: '1%',
                    height: '100%'
                }}
            >
                <Container fitHeight={true}>
                    <div>
                        <Box data-testid="tab-box">
                            <EntitiesList
                                entities={documentEntities}
                                documentPageCount={props.documentPageCount}
                                currentPageNumber={props.currentPageNumber}
                                switchPage={props.switchPage}
                                comprehendService={props.comprehendService}
                                entityType={props.entityType}
                                standardEntities={props.standardEntities}
                                medicalEntities={props.medicalEntities}
                                piiEntities={props.piiEntities}
                                selectedEntities={props.selectedEntities}
                                setSelectedEntities={props.setSelectedEntities}
                                selectedDocumentId={props.selectedDocumentId}
                                selectedCaseId={props.selectedCaseId}
                                previewRedaction={props.previewRedaction}
                                setPreviewRedaction={props.setPreviewRedaction}
                                currentStatus={props.currentStatus}
                                phrase={props.phrase}
                                setPhrase={props.setPhrase}
                                inputPhrase={inputPhrase}
                                setInputPhrase={setInputPhrase}
                            />
                        </Box>
                    </div>
                </Container>
            </div>
        </div>
    );
}
