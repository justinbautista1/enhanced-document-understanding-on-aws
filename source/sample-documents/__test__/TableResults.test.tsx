// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { StatusIndicatorProps } from '@cloudscape-design/components';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { tables } from './test_data';
import TableResults from '../components/TableResults/TableResults';

const tableResultsPropsEmpty = {
    currentPageNumber: 1,
    documentPageCount: 4,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'success' as StatusIndicatorProps.Type
};

const tableResultsProps = {
    currentPageNumber: 1,
    documentPageCount: 4,
    tables: tables,
    switchPage: jest.fn(),
    currentStatus: 'success' as StatusIndicatorProps.Type
};

const tableResultsPropsLoading = {
    currentPageNumber: 0,
    documentPageCount: 0,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'loading' as StatusIndicatorProps.Type
};

const tableResultsPropsFailed = {
    currentPageNumber: 0,
    documentPageCount: 0,
    tables: [],
    switchPage: jest.fn(),
    currentStatus: 'error' as StatusIndicatorProps.Type
};

test('Renders without tables', async () => {
    const view = render(<TableResults {...tableResultsPropsEmpty} />);
    expect(view.container).toMatchSnapshot();
    expect(screen.getByTestId('tables-nodata')).toBeInTheDocument();
    expect(screen.getByTestId('tables-nodata').textContent).toBe('No Tables detected');
});

test('Renders error message when currentStatus is error', async () => {
    render(<TableResults {...tableResultsPropsFailed} />);
    expect(screen.getByTestId('tables-status-only').textContent).toBe(' An error occurred loading table results. ');
});

test('Renders loading message when currentStatus is loading', async () => {
    render(<TableResults {...tableResultsPropsLoading} />);
    expect(screen.getByTestId('tables-status-only').textContent).toBe(' Loading ');
});

test('Renders tables with page numbers shown for multi-page', async () => {
    render(<TableResults {...tableResultsProps} />);
    expect(await screen.findByText('Tables: 2 Found')).toBeInTheDocument();
    expect(await screen.findByText('Page 1')).toBeInTheDocument();
    expect(await screen.findByText('Table 1 : 2 rows')).toBeInTheDocument();
    expect(await screen.findByText('Page 2')).toBeInTheDocument();
    expect(await screen.findByText('Table 1 : 1 rows')).toBeInTheDocument();
});

test('calls switchPage when a line is clicked', async () => {
    render(<TableResults {...tableResultsProps} />);
    const line = await screen.findByText('Table 1 : 1 rows');
    line.click();
    expect(tableResultsProps.switchPage).toHaveBeenCalledWith(2);
});
