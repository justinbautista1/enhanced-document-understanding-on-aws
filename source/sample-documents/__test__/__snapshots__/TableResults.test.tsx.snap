// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without tables 1`] = `
<div>
  <p
    data-testid="tables-nodata"
  >
    No Tables detected
  </p>
</div>
`;
